File: 1478-allocate-mailboxes/1478-allocate-mailboxes.js
/**
 * @param {number[]} houses
 * @param {number} k
 * @return {number}
 */
var minDistance = function(houses, k) {
        const dist = Array(houses.length).fill(0).map(house => Array(houses.length).fill(0))
        
        
        const n = houses.length
        houses.sort((a, b) => a -b)
        for(let i = 0; i< n; i++){
            for(let j = i; j < n; j++)
            {
                let mid = houses[Math.floor((i + j) / 2)]
                for(let h = i; h <= j; h++)
                    dist[i][j] += Math.abs(mid - houses[h])
        }
        }
    const dp = (k, i , cache) => {
            if (k == 0 && i == n) 
                return 0
            if (k == 0 || i == n) 
                return Infinity
           const key = `${k}-${i}`
           if (key in cache) return cache[key]
           let ans = Infinity
            for(let j = i; j < n; j++)
                ans = Math.min(ans, dist[i][j] + dp(k-1, j+1, cache))
            cache[key] = ans
            return ans
    }
    
    return dp(k, 0, {})
};

File: 1478-allocate-mailboxes/1478-allocate-mailboxes.py
class Solution:
    def minDistance(self, houses: List[int], k: int) -> int:
        
        lru_cache(None)
        def dp(k, i):
            if k == 0 and i == n :
                return 0
            if k == 0 or i == n :
                return sys.maxsize
            ans = sys.maxsize
            for j in range(i, n):
                ans = min(ans, dist[i][j] + dp(k-1, j+1))
            return ans
        dist = [[0] * len(houses) for _ in houses]
        
        
        n = len(houses)
        houses.sort()
        for i in range(n):
            for j in range(i, n):
                mid = houses[(i + j) // 2]
                for h in range(i, j+1):
                    dist[i][j] += abs(mid - houses[h])
        dp = [[sys.maxsize] * k for _ in houses]
        for i in range(n):
            dp[i][0] = dist[0][i]
        for kk in range(1, k):
            for i in range(0, n):
                for j in range(i):
                    dp[i][kk] = min(dp[i][kk] , dp[j][kk-1] + dist[j+1][i])
        return dp[-1][-1]